Added Multi Module data domain

kotrasatheesh-dev opened the pull request.

kotrasatheesh-dev committed:
Added Multi Module data domain

kotrasatheesh-dev committed:
Added Multi Module data domain

kotrasatheesh-dev committed:
Added Multi Module data domain

kotrasatheesh-dev committed:
Added Multi Module data domain

kotrasatheesh-dev committed:
Added Multi Module data domain

kotrasatheesh-dev committed:
Added Multi Module data domain

shreyansh-mm commented on common/src/main/java/common/module/ApolloModule.kt:
```diff
@@ -0,0 +1,21 @@
+        val apolloClient =
+            ApolloClient
+                .Builder()
+                .serverUrl("https://rickandmortyapi.com/graphql")
```
Keep them in a constant file

shreyansh-mm commented on common/src/main/java/common/module/ApolloModule.kt:
```diff
@@ -0,0 +1,21 @@
+package common.module
```
Move to data module

shreyansh-mm commented on data/src/main/java/data/repository/CharactersRepositoryImpl.kt:
```diff
@@ -0,0 +1,21 @@
+    private val apolloClient: ApolloClient,
+) : CharactersRepository {
+
+    override suspend fun getCharacters(): Flow<List<GetCharactersQuery.Result>> = flow {
```
1. Flow is not necessary
2. Use your own POJO instead of `GetCharactersQuery.Result`

shreyansh-mm commented on data/src/main/java/data/repository/CharactersRepositoryImpl.kt:
```diff
@@ -0,0 +1,21 @@
+    override suspend fun getCharacters(): Flow<List<GetCharactersQuery.Result>> = flow {
+        val response = apolloClient.query(GetCharactersQuery()).execute()
+        val characters = response.data?.characters?.results ?.filterNotNull()?: emptyList()
+        emit(characters)
```
Create custom exception class with different exceptions which can be mapped to all the exception which our repository can encounter and throw them instead.

kotrasatheesh-dev committed:
fix:PR comments
1.added Constants.kt file
2.custom data class Character.kt
3.CharacterMapper added conversion
4.updated graphql query
5.build.gradle added required dependency
6.repository necessary  changes

kotrasatheesh-dev committed:
fix:PR comments
1.removed files not needed

kotrasatheesh-dev committed:
fix:PR comments

kotrasatheesh-dev committed:
fix:PR comments

kotrasatheesh-dev committed:
fix:PR comments

kotrasatheesh-dev committed:
fix:PR comments

kotrasatheesh-dev committed:
feat:added test cases
1.CharacterUseCaseTest

kotrasatheesh-dev committed:
feat:added test cases
1.CharacterUseCaseTest

shreyansh-mm commented on data/build.gradle.kts:
```diff
@@ -33,11 +35,32 @@ android {
     testImplementation(libs.junit)
     androidTestImplementation(libs.androidx.junit)
     androidTestImplementation(libs.androidx.espresso.core)
+    testImplementation(libs.kotlinx.coroutines.test)
+    testImplementation(libs.mockk)
+    testImplementation(libs.androidx.core.testing)
+    testImplementation(libs.turbine)
+    testImplementation(libs.truth)
+    implementation(libs.apollo.runtime)
+    implementation(libs.dagger)
+    ksp(libs.dagger.compiler)
+    implementation(libs.androidx.paging.compose)
+    implementation(libs.androidx.paging.runtime)
```
Space out these dependencies.
And try to follow this order
api, implementation (ksp / kapt right after its needed dependency), testImplementation, androidTestImplementation

shreyansh-mm commented on data/src/main/java/data/module/CharacterMapper.kt:
```diff
@@ -0,0 +1,22 @@
+    object CharacterMapper {
+        fun map(result: GetCharactersQuery.Result): Character {
+            return Character(
+                id = result.id ?: "",
```
Why are we setting ID as empty string.

shreyansh-mm commented on data/src/main/java/data/module/CharacterMapper.kt:
```diff
@@ -0,0 +1,22 @@
+            return results
+                ?.filterNotNull()
+                ?.map { map(it) }
+                ?: emptyList()
```
Empty list should not be returned if input data is nullable.

shreyansh-mm commented on data/src/main/java/exception/RepositoryException.kt:
```diff
@@ -0,0 +1,7 @@
+package exception
+
+sealed class RepositoryException(override val message: String) : Exception(message) {
```
Please give this a better name

shreyansh-mm commented on domain/build.gradle.kts:
```diff
@@ -33,11 +36,21 @@ android {
     testImplementation(libs.junit)
     androidTestImplementation(libs.androidx.junit)
     androidTestImplementation(libs.androidx.espresso.core)
+    implementation(libs.apollo.runtime)
+    testImplementation(libs.kotlinx.coroutines.test)
+    testImplementation(libs.mockk)
+    testImplementation(libs.androidx.core.testing)
+    testImplementation(libs.turbine)
+    testImplementation(libs.truth)
+    implementation(libs.dagger)
+    ksp(libs.dagger.compiler)
+    implementation(libs.androidx.paging.compose)
+    implementation(libs.androidx.paging.runtime)
```
Same arrangement issue, please fix it.

shreyansh-mm reviewed (changes requested):
Tests for repository are missing

kotrasatheesh-dev committed:
Fix:PR comments fixed
1:data/build.gradle.kts
2:CharacterMapper.kt
3:domain/build.gradle.kts

shreyansh-mm commented on data/src/main/java/data/repository/CharactersRepositoryImpl.kt:
```diff
@@ -0,0 +1,35 @@
+                    errorMessage.toString()
+                )
+            }
+            return CharacterMapper.mapList(results)!!
```
We should not use !! operator

shreyansh-mm commented on data/src/main/java/data/module/CharacterMapper.kt:
```diff
@@ -0,0 +1,22 @@
+            )
+        }
+
+        fun mapList(results: List<GetCharactersQuery.Result?>?): List<Character>? {
```
Ideally we should not accept nullable input here or return nullable output

kotrasatheesh-dev committed:
Fix:PR comments resolved
